### PR TITLE
Add ColBERT embedding_mode for asymmetric query/document encoding

### LINKED ISSUE #42234: [Usage]: How to get query embeddings from ColBERT?
### Your current environment

```text
Collecting environment information...
==============================
        System Info
==============================
OS                           : Ubuntu 24.04.3 LTS (x86_64)
GCC version                  : (Ubuntu 13.3.0-6ubuntu2~24.04.1) 13.3.0
Clang version                : Could not collect
CMake version                : Could not collect
Libc version                 : glibc-2.39

==============================
       PyTorch Info
==============================
PyTorch version              : 2.11.0+cu130
Is debug build               : False
CUDA used to build PyTorch   : 13.0
ROCM used to build PyTorch   : N/A
XPU used to build PyTorch    : N/A

==============================
      Python Environment
==============================
Python version               : 3.12.3 (main, Mar 23 2026, 19:04:32) [GCC 13.3.0] (64-bit runtime)
Python platform              : Linux-6.8.0-110-generic-x86_64-with-glibc2.39
    
==============================
       CUDA / GPU Info
==============================
Is CUDA available            : True
CUDA runtime version         : 13.0.88
CUDA_MODULE_LOADING set to   : 
GPU models and configuration : GPU 0: NVIDIA GeForce GTX 1660 Ti
Nvidia driver version        : 595.58.03
cuDNN version                : Could not collect
HIP runtime version          : N/A
MIOpen runtime version       : N/A
Is XNNPACK available         : True

==============================
          CPU Info
==============================
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           39 bits physical, 48 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  8
On-line CPU(s) list:                     0-7
Vendor ID:                               GenuineIntel
Model name:                              Intel(R) Core(TM) i7-9700 CPU @ 3.00GHz
CPU family:                              6
Model:                                   158
Thread(s) per core:                      1
Core(s) per socket:                      8
Socket(s):                               1
Stepping:                                13
CPU(s) scaling MHz:                      95%
CPU max MHz:                             4700.0000
CPU min MHz:                             800.0000
BogoMIPS:                                6000.00
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid mpx rdseed adx smap clflushopt intel_pt xsaveopt xsavec xgetbv1 xsaves dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp vnmi md_clear flush_l1d arch_capabilities ibpb_exit_to_user
Virtualization:                          VT-x
L1d cache:                               256 KiB (8 instances)
L1i cache:                               256 KiB (8 instances)
L2 cache:                                2 MiB (8 instances)
L3 cache:                                12 MiB (1 instance)
NUMA node(s):                            1
NUMA node0 CPU(s):                       0-7
Vulnerability Gather data sampling:      Mitigation; Microcode
Vulnerability Indirect target selection: Mitigation; Aligned branch/return thunks
Vulnerability Itlb multihit:             KVM: Mitigation: VMX disabled
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Mitigation; Clear CPU buffers; SMT disabled
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Mitigation; Enhanced IBRS
Vulnerability Spec rstack overflow:      Not affected
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Enhanced / Automatic IBRS; IBPB conditional; PBRSB-eIBRS SW sequence; BHI SW loop, KVM SW loop
Vulnerability Srbds:                     Mitigation; Microcode
Vulnerability Tsa:                       Not affected
Vulnerability Tsx async abort:           Mitigation; TSX disabled
Vulnerability Vmscape:                   Mitigation; IBPB before exit to userspace

==============================
Versions of relevant libraries
==============================
[pip3] flashinfer-python==0.6.8.post1
[pip3] numpy==2.2.6
[pip3] nvidia-cublas==13.1.0.3
[pip3] nvidia-cuda-cupti==13.0.85
[pip3] nvidia-cuda-nvrtc==13.0.88
[pip3] nvidia-cuda-runtime==13.0.96
[pip3] nvidia-cudnn-cu13==9.19.0.56
[pip3] nvidia-cudnn-frontend==1.18.0
[pip3] nvidia-cufft==12.0.0.61
[pip3] nvidia-cufile==1.15.1.6
[pip3] nvidia-curand==10.4.0.35
[pip3] nvidia-cusolver==12.0.4.66
[pip3] nvidia-cusparse==12.6.3.3
[pip3] nvidia-cusparselt-cu13==0.8.0
[pip3] nvidia-cutlass-dsl==4.4.2
[pip3] nvidia-cutlass-dsl-libs-base==4.4.2
[pip3] nvidia-ml-py==13.595.45
[pip3] nvidia-nccl-cu13==2.28.9
[pip3] nvidia-nvjitlink==13.0.88
[pip3] nvidia-nvshmem-cu13==3.4.5
[pip3] nvidia-nvtx==13.0.85
[pip3] pyzmq==27.1.0
[pip3] torch==2.11.0+cu130
[pip3] torch_c_dlpack_ext==0.1.5
[pip3] torchaudio==2.11.0+cu130
[pip3] torchvision==0.26.0+cu130
[pip3] transformers==5.6.2
[pip3] triton==3.6.0
[conda] Could not collect

==============================
         vLLM Info
==============================
ROCM Version                 : Could not collect
vLLM Version                 : 0.20.0
vLLM Build Flags:
  CUDA Archs: 7.5 8.0 8.6 8.9 9.0 10.0 12.0+PTX; ROCm: Disabled; XPU: Disabled
GPU Topology:
  	GPU0	CPU Affinity	NUMA Affinity	GPU NUMA ID
GPU0	 X 	0-7	0		N/A

Legend:

  X    = Self
  SYS  = Connection traversing PCIe as well as the SMP interconnect between NUMA nodes (e.g., QPI/UPI)
  NODE = Connection traversing PCIe as well as the interconnect between PCIe Host Bridges within a NUMA node
  PHB  = Connection traversing PCIe as well as a PCIe Host Bridge (typically the CPU)
  PXB  = Connection traversing multiple PCIe bridges (without traversing the PCIe Host Bridge)
  PIX  = Connection traversing at most a single PCIe bridge
  NV#  = Connection traversing a bonded set of # NVLinks

==============================
     Environment Variables
==============================
NVIDIA_VISIBLE_DEVICES=all
NVIDIA_REQUIRE_CUDA=cuda>=13.0 brand=unknown,driver>=535,driver<536 brand=grid,driver>=535,driver<536 brand=tesla,driver>=535,driver<536 brand=nvidia,driver>=535,driver<536 brand=quadro,driver>=535,driver<536 brand=quadrortx,driver>=535,driver<536 brand=nvidiartx,driver>=535,driver<536 brand=vapps,driver>=535,driver<536 brand=vpc,driver>=535,driver<536 brand=vcs,driver>=535,driver<536 brand=vws,driver>=535,driver<536 brand=cloudgaming,driver>=535,driver<536 brand=unknown,driver>=550,driver<551 brand=grid,driver>=550,driver<551 brand=tesla,driver>=550,driver<551 brand=nvidia,driver>=550,driver<551 brand=quadro,driver>=550,driver<551 brand=quadrortx,driver>=550,driver<551 brand=nvidiartx,driver>=550,driver<551 brand=vapps,driver>=550,driver<551 brand=vpc,driver>=550,driver<551 brand=vcs,driver>=550,driver<551 brand=vws,driver>=550,driver<551 brand=cloudgaming,driver>=550,driver<551 brand=unknown,driver>=565,driver<566 brand=grid,driver>=565,driver<566 brand=tesla,driver>=565,driver<566 brand=nvidia,driver>=565,driver<566 brand=quadro,driver>=565,driver<566 brand=quadrortx,driver>=565,driver<566 brand=nvidiartx,driver>=565,driver<566 brand=vapps,driver>=565,driver<566 brand=vpc,driver>=565,driver<566 brand=vcs,driver>=565,driver<566 brand=vws,driver>=565,driver<566 brand=cloudgaming,driver>=565,driver<566 brand=unknown,driver>=570,driver<571 brand=grid,driver>=570,driver<571 brand=tesla,driver>=570,driver<571 brand=nvidia,driver>=570,driver<571 brand=quadro,driver>=570,driver<571 brand=quadrortx,driver>=570,driver<571 brand=nvidiartx,driver>=570,driver<571 brand=vapps,driver>=570,driver<571 brand=vpc,driver>=570,driver<571 brand=vcs,driver>=570,driver<571 brand=vws,driver>=570,driver<571 brand=cloudgaming,driver>=570,driver<571 brand=unknown,driver>=575,driver<576 brand=grid,driver>=575,driver<576 brand=tesla,driver>=575,driver<576 brand=nvidia,driver>=575,driver<576 brand=quadro,driver>=575,driver<576 brand=quadrortx,driver>=575,driver<576 brand=nvidiartx,driver>=575,driver<576 brand=vapps,driver>=575,driver<576 brand=vpc,driver>=575,driver<576 brand=vcs,driver>=575,driver<576 brand=vws,driver>=575,driver<576 brand=cloudgaming,driver>=575,driver<576
TORCH_CUDA_ARCH_LIST=7.5 8.0 8.6 8.9 9.0 10.0 12.0+PTX
NVIDIA_DRIVER_CAPABILITIES=compute,utility
VLLM_USAGE_SOURCE=production-docker-image
CUDA_VERSION=13.0.2
VLLM_ENABLE_CUDA_COMPATIBILITY=0
LD_LIBRARY_PATH=/usr/local/nvidia/lib64:/usr/local/cuda/lib64:/usr/local/nvidia/lib:/usr/local/nvidia/lib64:/usr/local/cuda/lib64
PYTORCH_NVML_BASED_CUDA_CHECK=1
TORCHINDUCTOR_COMPILE_THREADS=1
TORCHINDUCTOR_CACHE_DIR=/tmp/torchinductor_root
```

### How would you like to use vllm

I would like to create query embeddings with ColBERT Late Interaction Models. In the [docs](https://docs.vllm.ai/en/stable/models/pooling_models/specific_models/), there is no mention of a parameter with which you can toggle between creating query embeddings and document embeddings.

I am serving the `jinaai/jina-colbert-v2` model with this command:
```
docker run -d --runtime nvidia --gpus all -p 8000:8000 --ipc=host vllm/vllm-openai:v0.20.0-cu130-ubuntu2404 --model jinaai/jina-colbert-v2 --hf-overrides '{"architectures": ["ColBERTJinaRobertaModel"]}' --trust-remote-code
```

I tried adding the `input_type` parameter which i have found [here](https://docs.vllm.ai/en/latest/models/pooling_models/embed/#multi-modal-inputs).
```
curl -s http://localhost:8000/pooling -H "Content-Type: application/json" -d '{
    "model": "jinaai/jina-colbert-v2",
    "input": "What is machine learning?",
    "task": "token_embed",
    "input_type": "query"
}'
```
I also tried prepending `[Q]`
```
curl -s http://localhost:8000/pooling -H "Content-Type: application/json" -d '{
    "model": "jinaai/jina-colbert-v2",
    "input": "[Q] What is machine learning?",
    "task": "token_embed",
}'
```
and `[QueryMarker]`
```
curl -s http://localhost:8000/pooling -H "Content-Type: application/json" -d '{
    "model": "jinaai/jina-colbert-v2",
    "input": "[QueryMarker] What is machine learning?",
    "task": "token_embed",
}'
```

From my understanding, if the query embedding would have worked, the response would have consisted of 32 Vectors, no matter the length of the query. This was not the case for any of the examples above.

Any help would be greatly appreciated!

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by neerajrapelli**
## Summary

Adds `embedding_mode` (`query` | `document`) for ColBERT / late-interaction pooling models so query vs document tokenization happens at **pooling IO** (markers, query expansion, length limits), not in model `forward()`.

- Use **`POST /pooling`** with `task: "token_embed"` and top-level **`embedding_mode`**.
- Do **not** use `/v1/embeddings`, `pooling_params.embedding_mode`, or `input_type` for this flow.
- `/score` and `/rerank` continue to apply query/document encoding automatically via API fields.
- Omitting `embedding_mode` keeps the legacy symmetric `token_embed` path (backward compatible; not recommended for retrieval).

Fixes #42234

## Design

- Asymmetry at **text/tokenization** only (`ColBERTIOProcessorMixin`, `colbert_encoding.py`).
- `COLBERT_ATTENTION_MASK_ENFORCED_IN_FORWARD = False` — mask metadata only.
- Prefix-cache keys: canonical pre-tokenization text + model + mode + template/config hashes.
- Regression tests: `tests/models/pooling/test_colbert_embedding_mode.py`

## GPU validation (RunPod, `jinaai/jina-colbert-v2`, vLLM 0.22.0 + PR file overlay)

**Verdict:** READY TO MERGE — **19 PASS**, **0 FAIL**, **0 WARN**, **3 SKIP**

| Check | Result |
|-------|--------|
| Health | PASS |
| Pooling query / document / legacy | PASS |
| Score / rerank | PASS |
| Batch / cache / negatives / asymmetry | PASS |
| `pytest tests/models/pooling/test_colbert_embedding_mode.py` | **24/24 PASS** |

**Manual token counts** (`"machine learning"`):

| Mode | Tokens |
|------|--------|
| `embedding_mode: query` | 32 |
| `embedding_mode: document` | 5 |
| legacy (no mode) | 4 |

**Retrieval sanity:** query↔relevant doc **6.37** vs query↔irrelevant **2.41**.

### CODE REVIEW

File: vllm/inputs/engine.py [hunk 1]
```diff
@@ -25,6 +25,24 @@ class _InputOptions(TypedDict):
...
+
+    colbert_cache_key: NotRequired[str]
+    """Canonical ColBERT cache isolation key for this engine input."""
+
+    colbert_full_attention_mask: NotRequired[bool]
+    """When true, query expansion tokens use full attention (PyLate parity metadata)."""
+
+    colbert_attention_mask: NotRequired[list[int]]
+    """Per-token ColBERT attention mask for the model input builder."""
+
```
Comment (noooop) [inline vllm/inputs/engine.py:46]: Thanks for your contribution.

But in order to support ColBERT, too many core code changes were made. You know we can't possibly do this.

Comment (noooop) [reply vllm/inputs/engine.py:46]: Compared to ColBERT v2, ColBERT is not widely used, so we only plan to provide basic support—or even no support at all if it requires too much effort.

Try plugins. For reference:

https://github.com/vllm-project/vllm/tree/main/tests/plugins/bge_m3_sparse_plugin
https://github.com/vllm-project/vllm/blob/main/tests/plugins_tests/test_bge_m3_sparse_io_processor_plugins.py

Comment (neerajrapelli) [reply vllm/inputs/engine.py:46]: Okay my plan will be is to rework this into a plugin implementation and try to minimize other core changes. 

close this pr and move on to other issue? @noooop
